enhancements/installer/vsphere-ipi-zonal.md

```diff
@@ -279,24 +315,160 @@ type VSpherePlatformSpec struct {
+	DatacenterFailureDomain VSphereFailureDomainType = "Datacenter"
+)
+
+// VSpherePlatformFailureDomainSpec holds the region and zone failure domain and the vCenter topology of that failure domain.
+// +kubebuilder:validation:XValidation:rule="has(self.zoneType) && self.zoneType == 'HostGroup' ? self.topology.affinity.vmGroup != '' && self.topology.affinity.hostGroup != '' && self.topology.affinity.vmHostRule != '' : true",message="when zoneType is HostGroup, failuredomain topology affinity vmGroup, hostGroup and vmHostRule fields must be defined"
+// +kubebuilder:validation:XValidation:rule="has(self.topology.affinity) && (self.topology.affinity.vmGroup != '' || self.topology.affinity.hostGroup != '' || self.topology.affinity.vmHostRule != '') ? has(self.zoneType) && self.zoneType == 'HostGroup' : true",message="when failuredomain topology affinity vmGroup, hostGroup and vmHostRule fields are defined failuredomain zoneType must be HostGroup"
+type VSpherePlatformFailureDomainSpec struct {
+	// regionType defines the type of region that is used for this failure domain. The current available options
+	// are ComputeCluster and Datacenter.
+	// +openshift:validation:FeatureGateAwareEnum:featureGate=VSphereHostVMGroupZonal,enum=ComputeCluster;Datacenter
+	// +kubebuilder:validation:MinLength=9
+	// +kubebuilder:validation:MaxLength=14
+	// +kubebuilder:validation:Optional
+	// +optional
+	RegionType string `json:"regionType,omitempty"`
+
+	// zoneType defines the type of zone that is used for this failure domain. The current available options
+	// are HostGroup and ComputeCluster.
+	// +kubebuilder:validation:MinLength=9
+	// +kubebuilder:validation:MaxLength=14
+	// +kubebuilder:validation:Optional
+	// +openshift:validation:FeatureGateAwareEnum:featureGate=VSphereHostVMGroupZonal,enum=HostGroup;ComputeCluster
+	// +optional
+	ZoneType string `json:"zoneType,omitempty"`
+}
+
+// VSpherePlatformTopology holds the required and optional vCenter objects - datacenter,
+// computeCluster, networks, datastore and resourcePool - to provision virtual machines.
+type VSpherePlatformTopology struct {
+	// affinity holds the VMGroup and the HostGroup names in vCenter corresponds to 
+	// a vm-host group of type Virtual Machine and Host respectively. Is also
+	// contains the VMHostRule which is an affinity vm-host rule in vCenter.
+	// +openshift:validation:featureGate=VSphereHostVMGroupZonal
+	// +kubebuilder:validation:Optional
+	// +optional
+	Affinity VSphereFailureDomainAffinity `json:"affinity,omitempty"`
+}
+ 
+// VSphereFailureDomainAffinity contains the vCenter cluster vm-host group (virtual machine and host types) 
+// and the vm-host affinity rule that together creates a affinity configuration for vm-host based zonal. 
+// This configuration within vCenter creates the required association between a failure domain, virtual machines
+// and ESXi hosts to create a vm-host based zone.
+type VSphereFailureDomainAffinity struct {
```

Comment:
So, the thing - we will have to see is, if a customer specifies `computeCluster` as region and then specifies two different `hostGroup` as zones and assuming those `hostGroups` don't share the datastore - will CSI driver work...

Reply:
that scenario was just asked in slack the other day so yeah should be a concern.